ping/ping.c

```diff
@@ -758,47 +772,14 @@ int ping4_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 		close(probe_fd);
 
 	} else if (rts->device) {
-		struct sockaddr_in dst = rts->whereto;
-		struct ifreq ifr;
-		int fd = sock->fd;
-		int rc;
-		int errno_save;
-
-		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, rts->device, IFNAMSIZ - 1);
-
-		enable_capability_raw();
-		rc = setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, rts->device, strlen(rts->device) + 1);
-		errno_save = errno;
-		disable_capability_raw();
-
-		if (rc == -1) {
-			if (IN_MULTICAST(ntohl(dst.sin_addr.s_addr))) {
-				struct ip_mreqn imr;
-
-				if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0)
-					error(2, 0, "%s: %s", _("unknown interface"), rts->device);
-				memset(&imr, 0, sizeof(imr));
-				imr.imr_ifindex = ifr.ifr_ifindex;
-				if (setsockopt(fd, SOL_IP, IP_MULTICAST_IF,
-					       &imr, sizeof(imr)) == -1)
-					error(2, errno, "IP_MULTICAST_IF");
-			} else
-				error(2, errno_save, "SO_BINDTODEVICE %s", rts->device);
-		}
+		bind_to_device(rts, sock->fd, rts->whereto.sin_addr.s_addr);
 	}
 
 	if (rts->whereto.sin_addr.s_addr == 0)
 		rts->whereto.sin_addr.s_addr = rts->source.sin_addr.s_addr;
 
-	if (rts->device) {
-		struct ifreq ifr;
-
-		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, rts->device, IFNAMSIZ - 1);
-		if (ioctl(sock->fd, SIOCGIFINDEX, &ifr) < 0)
-			error(2, 0, _("unknown iface: %s"), rts->device);
-	}
+	if (rts->device)
+		iface_name2index(rts, sock->fd);
```

Comment:
Looking at this part of the code, is the result even used? Can't we just delete this part?

We get the result in the ifr.ifr_ifindex but before it was not used at all, now it's returned from the function but again the return value is not used.

Reply:
Here is used to check for unknown interface in case of ` -I <VRF interface> and -I <local IPv4 address>`:

```
$ ./builddir/ping/ping 1:2::3:4 -I nonexisting -I 1:2::3:4 -c 1
builddir/ping/ping: unknown iface: nonexisting
```
Maybe it's a bit confusing use the same function for both: 1) getting interface index 2) just for checking interface exists. Any idea how to do it better?

Also, This check should be moved to getopts, but I'd prefer to do this after release. `-I` can contain interface or address.

Reply:
Actually, you're right this check is probably not needed as it's used only for `I <VRF interface> and -I <local IPv4 address>`, but check for invalid device has been done earlier as part of `bind_to_device()` call in ping.c:786 `ping4_run()`.